### ISSUE TITLE
Skybell

### ISSUE THREAD

**Issue by lutzky** [uvjustin/alarmdotcom#122]
**Which device would you like us to support?**

Skybell HD: https://skybell.com/skybell-hd/

(Note: There's a separate homeassistant integration for skybell for those using it on skybell.com; we're using it with *alarm.com*)

**Where does this device appear in the Alarm.com mobile app?**

Video Devices (has a bell icon)

**What types of actions does this device support?**

The device is a combination camera (with two-directional audio) and doorbell; the camera is probably more complex to implement, and I'm more interested in the doorbell: It specifically supports turning the indoor and outdoor chime on and off, and the automation I'm looking to build is "turn the indoor chime off and remind me to turn it back on" - e.g. baby is napping.

**Which provider do you use?**

GoKonnect via alarm.com

**Include pyalarmdotcomajax Server Output**
This is an important step. It's helpful for us to see how this device is represented on the Alarm.com server. We have a tool for dumping this data via the command line.

1. Install [Python >= 3.9](https://www.python.org/downloads/).
2. Install [pyalarmdotcomajax](https://github.com/uvjustin/pyalarmdotcomajax) v0.2.0-beta via pip: `pip install pyalarmdotcomajax==0.2.0b0`.
3. Run `adc -u YOUR_USERNAME -p YOUR_PASSWORD -c YOUR_2FA_COOKIE -vx`
4. The above command dumps server data for all devices that are known to ADC developers, including a few devices that are currently unsupported. **Heads Up!** This command may leak sensitive information. It does _not_ dump your email address, address, or any other account information, but it may output coordinates from GPS devices (like an Alarm.com car sensor) and names of family members (e.g.: Michael's Room Window Sensor). Be sure to review the data before posting to scrub anything that you're not comfortable posting online.

Very odd output from this tool:

```
Provider is None
Logging in as lutzky.
DEBUG:pyalarmdotcomajax:Attempting to log in to Alarm.com
DEBUG:pyalarmdotcomajax:Response status from Alarm.com: 200
DEBUG:pyalarmdotcomajax:{'__VIEWSTATE': '[redacted]', '__VIEWSTATEGENERATOR': '[redacted]', '__EVENTVALIDATION': '[redacted]', '__PREVIOUSPAGE': '[redacted]'}
INFO:pyalarmdotcomajax:Attempting login to Alarm.com
DEBUG:pyalarmdotcomajax:Got Provider: GoKonnect Smart Home Security, User ID: [redacted]
DEBUG:pyalarmdotcomajax:Calling update on Alarm.com
DEBUG:pyalarmdotcomajax.entities:Initialized ADCRelationshipType.SYSTEM (systems/system) Lutzky's home
DEBUG:pyalarmdotcomajax.entities:Initialized ADCRelationshipType.PARTITION (devices/partition) Panel
DEBUG:pyalarmdotcomajax.entities:Initialized ADCRelationshipType.SENSOR (devices/sensor) [redacted]
[...several more of these, with correct names, as they appear in homeassistant too...]
DEBUG:pyalarmdotcomajax.entities:Initialized ADCRelationshipType.SENSOR (devices/sensor) [redacted]
DEBUG:pyalarmdotcomajax:Failed to get data. Response: [{'status': '423', 'detail': 'You do not have permissions for this device type', 'code': 423}]
Permission error. Check that your credentials are correct.
```

Happy to try and debug this further :shrug: 

**Include Action Endpoints**

From using HTTP Toolkit with the web UI, looks like it's POSTing a request to `https://www.alarm.com/web/Video/SettingsMain_V2.aspx` with a fairly opaque 42.9kB request body. It does contain stuff like `ctl00%24phBody%24cbIndoorChime=on&`. I really hope there's another way around this, it looks like a pain to reverse-engineer.

**Comment by elahd** [comment]
Hey @lutzky. Thanks for the detail. I fixed the bug the caused the crash on the 423 above. Can you update pyalarmdotcomajax to v0.2.10 and try running the same command again? (Just the `-vx` flags, not `-d`.) Update via `pip install pyalarmdotcomajax==0.2.10`.

You're right that this will be a tough device to onboard. Alarm.com's APIs for device configs aren't RESTful.

Would you mind posting a screenshot of the Alarm.com config page for the bell? I may have an easier time poking around if I can see the available options

**Comment by lutzky** [comment]
Thanks for taking a look!

adc update seems to have fixed the issue; here's the relevant snippet:

```
====[Camera]====                                                                 
                                                                                 
{"data": [{"id": "REDACTED", "type": "video/camera", "attributes": {"supportsSnapShot": false, "canTakeSnapshot": false, "canRequestRecording": false, "canUploadClips": true, "hasSvrSchedule": false, "excludedFromVisualVerification": false, "liveViewHdToggleEnabled": true, "supports4MPStreaming": false, "canPlaySiren": false, "isSirenEnabled": false, "isSirenOn": false, "canPanTilt": false, "maxPanTiltLevel": 0, "canZoom": false, "maxZoomLevel": 0, "minZoomLevelForPanTilt": 0, "panTiltPresets": [], "preferredPanTiltPreset": null, "isUnreachable": false, "lastSuccessfulSupervision": null, "canStreamToPanel": true, "canChangeStreamToPanel": false, "canStreamToPanelBeEnabled": true, "isStreamToPanelEnabled": false, "doorbellChimeType": 0, "isBatteryCameraAndUsingGatewayForBackendAccess": false, "supportsVmdSchedules": false, "deviceImagePath": "/web/webimages/camera_images/SkyBellHD_64px.png?v=2", "deviceModel": "SKYBELLHD", "macAddress": "REDACTED", "firmwareVersion": "REDACTED", "publicIp": null, "privateIp": null, "port": null, "httpsPort": null, "shouldUseEntireImageForSnapshot": false, "hasDdnsed": false, "usageProtocolMapping": {"18": 3, "10": 1, "6": 1}, "isVirtualCamera": false, "isVarifocalCamera": false, "hasSdRecordingSchedule": false, "supportsLiveView": false, "forcedAspectRatio": null, "canBeSaved": false, "canChangeDescription": false, "description": "Front door", "deviceModelId": 0, "canConfirmStateChange": false, "canReceiveCommands": false, "remoteCommandsEnabled": false, "hasPermissionToChangeState": false, "deviceIcon": {"icon": 27}, "batteryLevelNull": null, "lowBattery": false, "criticalBattery": false, "canBeLinkedToVideoDevice": false, "videoDeviceLinkDeviceIds": null}, "relationships": {"videoEvents": {"data": [], "meta": {"count": "0"}}, "videoSource": {"data": {"id": "REDACTED", "type": "video/liveVideoSource"}}, "videoSourceHd": {"data": {"id": "REDACTED", "type": "video/liveVideoHighestResSource"}}, "svrVideoSource": {"data": {"id": "REDACTED", "type": "video/videoSources/svrVideoSource"}}, "svrScrubberThumbnailEndpoint": {"data": {"id": "96747708-2048", "type": "video/svrThumbnailEndpoint"}}, "cameraSiren": {"data": null}, "privacyZonesSetting": {"data": null}, "snapshot": {"data": {"id": "REDACTED", "type": "video/snapshot"}}, "system": {"data": {"id": "REDACTED", "type": "systems/system"}}, "stateInfo": {"data": null}}}], "included": [], "meta": {"transformer_version": "1.1"}}
```
![skybell](https://user-images.githubusercontent.com/8140/165390262-b25f2688-cbbc-4b4c-a900-100d16fb52de.png)

Let me know if you need some of the IDs unredacted. This seems odd, it doesn't quite match.

**Comment by elahd** [comment]
Thanks. My login isn't allowing me to access anything within the /web/Video path. The adc dump returns device state info, not settings. It looks like Alarm.com considers the "Chime Type" options to be in the settings realm; not state. 

It would help to have:

1. The entire payload that contains `ctl00%24phBody%24cbIndoorChime=on&` (redacted is fine).
2. The URL of the settings page (screenshotted above).
3. The source of the settings page (redacted is fine).

**Comment by lutzky** [comment]
Not yet that משוגעס, but we'll see :sweat_smile: 

The URL of the screenshotted settings page is https://www.alarm.com/web/Video/SettingsMain_V2.aspx; yes, the same URL that is POSTed to is the URL the browser is showing.

- Request body payload: [reqbody.txt](https://github.com/uvjustin/alarmdotcom/files/8594988/reqbody.txt)
- Same, but URLdecoded (helpfully suggested by HTTP Toolkit): [reqbody_urldecoded.txt](https://github.com/uvjustin/alarmdotcom/files/8594990/reqbody_urldecoded.txt)
- Source of settings page (with `txt` extension because github is picky): [source.html.txt](https://github.com/uvjustin/alarmdotcom/files/8594993/source.html.txt)
.

**Comment by elahd** [comment]
I hacked away at this and have basic read functions ready. Try this on your personal computer:

1. Uninstall pyalarmdotcomajax: `pip uninstall pyalarmdotcomajax`.
2. Install the v3 dev branch: `pip install git+https://github.com/uvjustin/pyalarmdotcomajax@tests_extensions_skybell`.
3. Ensure that install was successful: `adc --version` should return `adc 0.3.0-dev`.
4. Use adc to pull your devices: `adc -u YOUR_USERNAME -p YOUR_PASSWORD`. Cameras will appear at the very bottom. You want to see something like this in the cameras section:

```
Front Doorbell (555555555-90)
ATTRIBUTES: [BATTERY: Normal] [READ ONLY: True] 
SETTINGS: {{ Indoor Chime [Type: Chime] [State: True] }} {{ Outdoor Chime [Type: Chime] [State: False] }} 
```

"Settings" is the new feature. If that's populated for you, then adc can successfully read state. Try changing the values through the Alarm.com website and see if they update in adc.

Once you do that, please report back. Next step is to test writing state through adc, followed by rolling this into the Home Assistant integration.

**Comment by elahd** [comment]
Thanks. The regular `pip install pyalarmdotcomajax` command installs the current production version of pyalarmdotcomajax. That version knows how to retrieve basic camera metadata but doesn't have any code to pull camera configuration data, which includes settings for chimes.

The dev version is failing on the new configuration retrieval code. I think that I fixed the issue. Either way, I added more debug code around the point of failure so that we get a more detailed error message if it fails again.

Can you follow the same steps that you ran through before? This time, you should expect to see `adc --version` return `0.3.0-dev.1`.

If it fails again in the same spot, please post the following (as well as any exceptions you see):

1) You should see an error message like this:

```
pyalarmdotcomajax.extensions:extensions.py:379 Unable to extract field. Failed on field ctl00$phBody$bridgeInfo$wirelessSettings$ctl07.
```

2. Re-run the command with the `-d` flag: `adc -u YOUR_USERNAME -p YOUR_PASSWORD -d`. You should see a full dump of the HTTP response for the configuration page bracked by `====== HTTP DUMP BEGIN ======` and `====== HTTP DUMP END ======`.

**Comment by lutzky** [comment]
```
root@c4539bf359d4:/# adc --version
adc 0.3.0-dev.1
root@c4539bf359d4:/# adc -u lutzky -p "$PASSWORD"
Logging in as lutzky.
ERROR:pyalarmdotcomajax.extensions:Unable to extract field. Failed on field __EVENTTARGET.
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/helpers.py", line 40, in extract_field_value
    if field.attrs.get("name") and field.name == "select":
AttributeError: 'NoneType' object has no attribute 'attrs'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/adc", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/cli.py", line 374, in main
    asyncio.run(cli())
  File "/usr/local/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/cli.py", line 160, in cli
    await alarm.async_update()
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/__init__.py", line 342, in async_update
    raise err
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/__init__.py", line 339, in async_update
    await self._async_get_and_build_devices(device_types)
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/__init__.py", line 642, in _async_get_and_build_devices
    ] = await extension_controller.fetch(
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/extensions.py", line 186, in fetch
    current_form_data = self._extract_fields(camera_config_id, tree)
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/extensions.py", line 382, in _extract_fields
    raise err
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/extensions.py", line 357, in _extract_fields
    value = extract_field_value(field)
  File "/usr/local/lib/python3.10/site-packages/pyalarmdotcomajax/helpers.py", line 50, in extract_field_value
    raise UnexpectedDataStructure from err
pyalarmdotcomajax.errors.UnexpectedDataStructure
```

So, no, the "Unable to extract field" message doesn't seem to show up. It would appear that something is calling `extract_field_value(None)`... perhaps `tree.find(attrs={"name": field_name})` is returning None? Probably worth guarding against that?

The `-d` output does not include the HTTP response output: https://gist.github.com/lutzky/86422181c743022caafdb7ff0ba55691

**Comment by elahd** [comment]
I'm curious as to why `__EVENTTARGET` is missing so that I know whether to ignore the error or fail. `0.3.0-dev.2` dumps the HTTP response body to the console even without a failure, so that'll help understand the cause. It's possible that some part of the request is malformed and Alarm.com is returning an error page.

Can you try again with the new version?

**Comment by lutzky** [comment]
https://gist.github.com/lutzky/45832ac85cc5b43f05117a1b5320c901

**Comment by elahd** [comment]
Try once more with v0.3.0-dev.3.

**Comment by lutzky** [comment]
Yay, it appears to work now!

With the chime on:

```
*** CAMERAS ***

--------
Front door (967**REDACTED**48)
ATTRIBUTES: [BATTERY: Normal] [READ ONLY: True] 
SETTINGS: {{ Indoor Chime [Type: Chime] [State: True] }} {{ Outdoor Chime [Type: Chime] [State: True] }} 

~~MALFUNCTION~~

--------
```

With the chime off, the respective `State` field changes from `True` to `None` (not `False`).

**Comment by lutzky** [comment]
I don't know what `~~MALFUNCTION~~` is indicating though.

**Comment by elahd** [comment]
Alright, let's test changing the chimes. Download v0.3.0-dev.4.

The command line syntax changed a bit. The `adc -u lutzky -p "$PASSWORD"` command you used before to pull data is now `adc -u lutzky -p "$PASSWORD" get`.

To change the chime, use:
```
adc -u lutzky -p "$PASSWORD" set -i DEVICE_ID -s SETTING_SLUG -k NEW_VALUE
```
Where:

1. DEVICE_ID is the Alarm.com device ID for your Skybell (should vaguely look like 108420342-1935).
2. SETTING_SLUG is either `indoor-chime-on` or `outdoor-chime-on`.
3. NEW_VALUE is either `True` or `False`.

E.g.: `adc -u lutzky -p "$PASSWORD" set -i 108420342-1935 -s indoor-chime-on -k False` turns off the indoor chime for 108420342-1935.

**Comment by lutzky** [comment]
No dice; claims to work and doesn't.

The `~~MALFUNCTION~~` string is gone, so that's good. 

Running `get` gets me (number swapped for the one in your example):

```
Front door (108420342-1935)                                                                                                                  
ATTRIBUTES: [BATTERY: Normal] [READ ONLY: True] 
SETTINGS: {{ Indoor Chime (indoor_chime_on) [Type: Chime] [State: True] }} {{ Outdoor Chime (outdoor_chime_on) [Type: Chime] [State: True] }
}                                                                     
```

The slug `indoor-chime-on` doesn't work; my guess of `indoor_chime_on` instead was correct, i.e. matches the `SETTINGS` stanza.

However, none of the options actually does anything - `{in,out}door_chime_on -k {True,False}`, none of the 4 options has any effect; neither in the alarm.com UI nor in the output of `get`. The output message is a green `Indoor Chime was successfully changed to False for Front door`.

Output with `-d`: https://gist.github.com/lutzky/1fcde9cdf23f25d4693077b6284236a4

**Comment by elahd** [comment]
Thanks. Can you try once more with -d on 0.3.0-dev.5?

**Comment by lutzky** [comment]
https://gist.github.com/lutzky/3c4724d51d0773ea57bb85327f2af00d

**Comment by elahd** [comment]
Thanks. I think we're almost there. There was a CLI bug that converted both True and False to True when processing new setting values 🤦🏻‍♂️. Should be fixed now. Can you test again with dev.6?

**Comment by lutzky** [comment]
With dev.6: https://gist.github.com/lutzky/489e9286f2c8334be07cc1139a7d8edc

**Comment by elahd** [comment]
Once more with dev.7. Python's aiohttp library serializes None types into literal "None" values. Now this library will manually convert None -> "" before POSTing.

**Comment by lutzky** [comment]
:shrug: 

In the interest of keeping things organized for myself, I'll just start updating the same gist: https://gist.github.com/lutzky/489e9286f2c8334be07cc1139a7d8edc

**Comment by elahd** [comment]
Again for dev.8, plz.

**Comment by elahd** [comment]
dev.9

**Comment by lutzky** [comment]
Back to previous failure now; updated same gist.

**Comment by elahd** [comment]
Released in v0.2.8.
